File: MATLAB Simulations/adc02revised/MEKF.m
classdef MEKF
    
    properties
        R % measurement covariance 
        first_estimate_done % A boolean
        P_update % meaurement updated state cov
        P_pre % predicted state cov
        P_prop % propagated state cov
        q_prop % propagated quaternion of body frame
        q_est % estimated quaternion
        del_x % error/incremental in state vector
        x_ref % reference state vector
        x_pred % predicted state
        x_prop % propagated state
        x_est % estimated state
        q_ref % reference quaternion
        predicted_meas % measurement prediction from propagated state
        dt_prop % propagation time between state predictions
        std_dev_gyro; % 6-D vector, first three components are std-dev in gyro bias measurements, next three are std dev in the gyro noise measurements
    end

    methods
        function obj = MEKF(R_input, P_start, x_init, dt, std_dev_process)
            obj.R = R_input;
            obj.q_prop = [0; 0; 0; 1];
            obj.q_ref = [0; 0; 0; 1];
            obj.q_est = [0; 0; 0; 1];
            obj.del_x = zeros(6, 1);
            obj.P_pre = P_start;
            obj.P_update = P_start;
            obj.P_prop = P_start;
            obj.x_ref = x_init;
            obj.dt_prop = dt;
            obj.x_prop = x_init; 
            %obj.predicted_meas = pred_measurement(obj.q_ref, [1; 0; 0], [0.1234; 0.4567; 0.7890]);
            obj.std_dev_gyro = std_dev_process;
            obj.first_estimate_done = false;
        end

        function obj = reset(obj)
            % Resets the x vector after setting dx vector to zero,
            % updates the reference quaternion 
            eps = epsilon(obj.q_ref);
            del_theta = obj.del_x(1:3); % the incremental angular dispacement vector from the reference quaternion

            obj.q_ref = obj.q_ref + 0.5 * eps * del_theta;
            obj.q_ref = obj.q_ref / norm(obj.q_ref); % normalising the quaternion;

            obj.x_ref(4:6) = obj.x_ref(4:6) + obj.del_x(4:6);
            obj.del_x = zeros(6, 1);
        end

        function obj = measurement_update(obj, zk, sun_prop_iner, mag_prop_iner) % zk is the measurement
            % Taking the measurement as sun vector followed by magnetic field vector
            if obj.first_estimate_done == true
                
            

            % calculating xk(+)
                Rk = obj.R;
                sun_prop_iner = sun_prop_iner/norm(sun_prop_iner);
                mag_prop_iner = mag_prop_iner/norm(mag_prop_iner);
                hxk = pred_measurement(obj.q_prop, sun_prop_iner, mag_prop_iner);
    
                Hk = sensitivity_matrix(hxk);
                P_pro = obj.P_prop;
                
                S = Hk * P_pro * Hk' + Rk;
                K = (P_pro * Hk') / S; % Kalman Gain
                
                zk = [zk(1:3)/norm(zk(1:3)); zk(4:6)/norm(zk(4:6))];
                obj.del_x = obj.del_x + K * (zk - hxk);
                
                obj.x_prop = obj.x_ref + obj.del_x;
                obj.P_update = (eye(6) - K * Hk) * obj.P_prop;
    
                obj.P_prop  = obj.P_update;
                
                K_del_z = K * (zk - hxk);
                del_theta= K_del_z(1:3);% the incremental angular dispacement vector from the reference quaternion
                obj.q_est = obj.q_prop + 0.5 * epsilon(obj.q_prop) * del_theta;
                obj.q_est = obj.q_est/norm(obj.q_est);
    
                obj.q_prop = obj.q_est;
            
             
                else 
                 
                b_meas_first = zk(4:6);
                sun_meas_first = zk(1:3);
                
                b_prop_first = mag_prop_iner;
                sun_prop_first = sun_prop_iner;
                
                v1 = b_prop_first;
                v1 = v1/norm(v1);
                v2 = cross(v1, sun_prop_first);
                v2 = v2/norm(v2);
                v3 = cross(v1, v2);
                V = [v1, v2, v3];

                w1 = b_meas_first;
                w1 = w1/norm(w1);
                w2 = cross(w1, sun_meas_first);
                w2 = w2/norm(w2);
                w3 = cross(w1, w2);
                W = [w1, w2, w3];
                
                A_start = W * V';
                q_start = quaternion_from_attitude(A_start);
                

                obj.q_prop = q_start;
                obj.q_ref = q_start;
                obj.q_est = q_start;
                obj.first_estimate_done = true;
                
            end
            
        end

        function obj = predict_state(obj, omega_meas)
            % Takes the angular velocity measurement and carries out state prediction
            % Calculating xk(-)
            w = omega_meas - obj.x_prop(4:6)  % subtracting the bias
            omega_meas = omega_meas
            temp= obj.x_prop(4:6)
            I3 = eye(3);
            o3 = zeros(3,3);
            F = [-skew(w), -I3; o3, o3];
            I6 = eye(6);
            phi = I6 + F * obj.dt_prop; % state transition matrix
            
            obj.x_prop = obj.x_prop + F * obj.dt_prop * obj.del_x;
            obj.del_x = phi * obj.del_x;
            

            % Qt calculation still has to be confirmed
            sigma_gyro_bias = diag([obj.std_dev_gyro(1)^2; obj.std_dev_gyro(2)^2; obj.std_dev_gyro(3)^2]);
            sigma_gyro_noise = diag([obj.std_dev_gyro(4)^2; obj.std_dev_gyro(5)^2; obj.std_dev_gyro(6)^2]);

            dt = obj.dt_prop;
            Qt = [sigma_gyro_bias * dt + 1/3 * sigma_gyro_noise * dt ^ 3, - sigma_gyro_noise * dt * dt/2; - sigma_gyro_noise * dt * dt/2, sigma_gyro_noise * dt ];
            obj.P_pre = phi * obj.P_pre * phi' + Qt; 

            % del_theta = obj.del_x(1:3); % the incremental angular dispacement vector from the reference quaternion
            obj.q_est = obj.q_prop + 0.5 * epsilon(obj.q_prop) * w * obj.dt_prop;
            obj.q_est = obj.q_est/norm(obj.q_est);

            

            obj.q_prop = obj.q_est;
        end
    end
end



function a = attitude_matrix(q)
    % Returns the attitude matrix of transformation from inertial to body frame

    q1 = q(1);
    q2 = q(2);
    q3 = q(3);
    q4 = q(4); % q4 is the scalar component

    a = [
        1 - 2*(q2^2 + q3^2), 2*(q1*q2 + q3*q4), 2*(q1*q3 - q2*q4);
        2*(q1*q2 - q3*q4), 1 - 2*(q1^2 + q3^2), 2*(q2*q3 + q1*q4);
        2*(q1*q3 + q2*q4), 2*(q2*q3 - q1*q4), 1 - 2*(q1^2 + q2^2)
    ];
end

function H = sensitivity_matrix(meas_pre)
% takes in the predicted sun-vector and magnetic field vector to return H
    b_pred = meas_pre(4:6);
    sun_pred = meas_pre(1:3);

    sun_skew = skew(sun_pred);
    b_skew = skew(b_pred);

    o3 = zeros(3,3);

    H = [sun_skew, o3; b_skew, o3];
end

function a = epsilon(q)
    % a helper matrix for quaternion multiplication
    a = [q(4)*eye(3) + skew(q(1:3));
         -q(1:3)'];
end

function a = skew(x)
    % returns the cross product matrix after taking in a 3 * 1 vector
    a = [0 -x(3) x(2);
         x(3) 0 -x(1);
         -x(2) x(1) 0];
end

function res = pred_measurement(q_temp, sun_prop_inertial, mag_prop_inertial)
    % Returns the body frame measurement of the sun-vector and the magnetic-field vector
    A = attitude_matrix(q_temp);
    b_rotated = A * mag_prop_inertial;
    sun_vec_rotated = A * sun_prop_inertial;
    res = zeros(6,1);
    res(1:3) = sun_vec_rotated;
    res(4:6) = b_rotated;
end

function q = quaternion_from_attitude(M)
    % Ensure M is a 3x3 matrix
    assert(all(size(M) == [3 3]), 'Input matrix must be 3x3');

    % Calculate trace of the matrix
    tr = M(1,1) + M(2,2) + M(3,3);

    if tr > 0
        % Case 1: Trace is positive
        S = sqrt(tr + 1.0) * 2; % S = 4 * qw
        qw = 0.25 * S;
        qx = (-M(3,2) + M(2,3)) / S;
        qy = (-M(1,3) + M(3,1)) / S;
        qz = (-M(2,1) + M(1,2)) / S;
    elseif M(1,1) > M(2,2) && M(1,1) > M(3,3)
        % Case 2: m00 is the largest diagonal element
        S = sqrt(1.0 + M(1,1) - M(2,2) - M(3,3)) * 2; % S = 4 * qx
        qw = (-M(3,2) + M(2,3)) / S;
        qx = 0.25 * S;
        qy = (M(1,2) + M(2,1)) / S;
        qz = (M(1,3) + M(3,1)) / S;
    elseif M(2,2) > M(3,3)
        % Case 3: m11 is the largest diagonal element
        S = sqrt(1.0 + M(2,2) - M(1,1) - M(3,3)) * 2; % S = 4 * qy
        qw = (-M(1,3) + M(3,1)) / S;
        qx = (M(1,2) + M(2,1)) / S;
        qy = 0.25 * S;
        qz = (M(2,3) + M(3,2)) / S;
    else
        % Case 4: m22 is the largest diagonal element
        S = sqrt(1.0 + M(3,3) - M(1,1) - M(2,2)) * 2; % S = 4 * qz
        qw = (-M(2,1) + M(1,2)) / S;
        qx = (M(1,3) + M(3,1)) / S;
        qy = (M(2,3) + M(3,2)) / S;
        qz = 0.25 * S;
    end

    % Return the quaternion as a column vector
    q = [qx; qy; qz; qw];
end

    % Combine into a quaternion with the scalar part as the last component
    

File: MEKF_Pico/mekf.py
from matrix import Matrix
from vector import Vector
from quaternion import Quaternion, quaternion_from_attitude_matrix

def sensitivity_matrix(pred_meas):
    ## pred_meas is expected to contain the unit vectors along measured acceleration and measured magnetic field
    assert(isinstance(pred_meas, Vector) and pred_meas.shape == (3,1), 'acc_prop inner should be a 3 * 1 vector')

    g_pred = Vector(pred_meas[0], pred_meas[1], pred_meas[2])
    b_pred = Vector(pred_meas[3], pred_meas[4], pred_meas[5])

    g_skew = g_pred.cross_pdt_matrix
    b_skew = b_pred.cross_pdt_matrix
    
    o3 = Matrix.zeros(3,3)

    H = Matrix(6,6)

    for i in range(3):
        for j in range(3):
            H[i][j] = g_skew[i][j]
            H[3 + i][j] = b_skew[i][j]
            H[3 + i][3 + j] = o3[i][j]
            H[i][3 + j] = o3[i][j]
    return H

def pred_measurement(q_curr, acc_prop_iner, mag_prop_iner):
    assert(isinstance(q_curr, Quaternion), 'First argument passed to pred_measurement should be a Quaternion')
    A = q_curr.attitude_matrix()

    assert(isinstance(acc_prop_iner, Vector) and acc_prop_iner.shape == (3,1), 'acc_prop inner should be a 3 * 1 vector')
    assert(isinstance(mag_prop_iner, Vector) and mag_prop_iner.shape == (3,1), 'mag_prop inner should be a 3 * 1 vector')

    b_rotated = A * mag_prop_iner
    g_rotated = A * acc_prop_iner

    hxk = Vector(g_rotated[0], g_rotated[1], g_rotated[2], b_rotated[0], b_rotated[1], b_rotated[2])

    return hxk

class MEKF:

    def __init__(self, R_input, P_start, x_init, inertial_acc_mag, dt, std_dev_process):
        self.R = R_input
        self.P_prop = P_start
        self.P_pre = P_start
        self.first_estimate_done = False
        self.q_est = Quaternion(1,0,0,0)
        self.q_prop = Quaternion(1,0,0,0)
        self.q_ref = Quaternion(1,0,0,0)
        self.del_x = Vector(0,0,0,0,0,0)
        self.x_prop = x_init
        self.x_est = x_init
        self.x_ref = x_init
        self.dt_prop = dt
        self.std_dev_gyro = std_dev_process
        self.z_reference = inertial_acc_mag

    def reset(self):
        eps = self.q_est.epsilon()
        del_theta = Vector(self.del_x[0], self.del_x[1], self.del_x[2])

        self.q_ref = self.q_ref + 0.5 * eps * del_theta
        self.q_ref.normalize()

        # Bias update
        for i in range(3,6):
            self.x_ref[i] = self.x_ref[i] + self.del_x[i]
        
        self.del_x = Vector(0,0,0,0,0,0)

    def measurement_update(self, acc_measure, mag_measure, inertial_propagation = -1):
        if inertial_propagation == -1:
            g_inertial = Vector(self.z_reference[0], self.z_reference[1], self.z_reference[2])
            b_inertial = Vector(self.z_reference[3], self.z_reference[4], self.z_reference[5])
        else:
            g_inertial = Vector(inertial_propagation[0], inertial_propagation[1], inertial_propagation[2])
            b_inertial = Vector(inertial_propagation[3], inertial_propagation[4], inertial_propagation[5])    
        
        b_meas = mag_measure.unit_vector
        g_meas = acc_measure.unit_vector
        ## Has to be completed
        if self.first_estimate_done:
            Rk = self.R

            hxk = pred_measurement(self.q_prop, g_inertial, b_inertial)

            H = sensitivity_matrix(hxk)

            P_pro = self.P_prop
            H_T = H.transpose()
            S = H * P_pro * H_T + Rk
            K = (P_pro * H_T) * S.inverse()

            zk = Vector(acc_measure[0], acc_measure[1], acc_measure[2], mag_measure[0], mag_measure[1], mag_measure[2])
            K_del_z = K * (zk - hxk)
            self.del_x = self.del_x + K_del_z

            del_theta = Vector(K_del_z[0], K_del_z[1], K_del_z[2])
            self.q_est = self.q_prop + 0.5 * self.q_prop.epsilon() * del_theta
            self.q_est.normalize()

            I6 = Matrix.identity(6)

            self.x_prop = self.x_ref + self.del_x
            self.P_prop = (I6 - K * H) * self.P_prop

            self.q_prop = self.q_est

        else:
            ## Applying TRIAD
            v1 = g_inertial
            v2 = v1.cross_pdt_matrix * b_inertial
            v2.normalize()
            v3 = v1.cross_pdt_matrix * v2

            V = Matrix.from_list([v1, v2, v3]).transpose()

            w1 = g_meas
            w2 = w1.cross_pdt_matrix * b_meas
            w2.normalize()
            w3 = w1.cross_pdt_matrix * w2

            W = Matrix.from_list([w1, w2, w3]).transpose()

            A_start = W * V.transpose()

            q_start = quaternion_from_attitude_matrix(A_start)
            self.q_est = q_start
            self.q_prop = q_start
            self.q_ref = q_start
            self.first_estimate_done = True
        
    def predict_state(self, omega_meas):

        bias = Vector(self.x_prop[0], self.x_prop[1], self.x_prop[2])
        w = omega_meas - bias
        dt = self.dt_prop
        w_cross = w.cross_pdt_matrix

        F = Matrix.zeros(6,6)

        I3 = Matrix.identity(3)

        for i in range(3):
            for j in range(3):
                F[i][j] = - w_cross[i][j]
                F[i][j + 3] = - I3[i][j]
        
        I6 = Matrix.identity(6)
        phi = I6 + F * dt

        self.x_prop = self.x_prop + F * self.del_x * dt
        self.del_x = phi * self.del_x

        sigma_gyro_bias = Matrix.diagonal(self.std_dev_gyro[0], self.std_dev_gyro[1], self.std_dev_gyro[2])
        sigma_gyro_noise = Matrix.diagonal(self.std_dev_gyro[3], self.std_dev_gyro[4], self.std_dev_gyro[5])
        sigma_gyro_bias = sigma_gyro_bias * sigma_gyro_bias
        sigma_gyro_noise = sigma_gyro_noise * sigma_gyro_noise 
        
        Q1 = sigma_gyro_bias * dt + (1/3 * sigma_gyro_noise * (dt ** 3))
        Q2 = - 0.5 * sigma_gyro_noise * (dt ** 2) 
        Q4 = sigma_gyro_noise * dt

        Q_k = Matrix(6,6)

        for i in range(3):
            for j in range(3):
                Q_k[i][j] = Q1[i][j]
                Q_k[3 + i][j] = Q2[i][j]
                Q_k[i][3 + j] = Q2[i][j]
                Q_k[3 + i][3 + j] = Q4[i][j]

        self.P_pre = phi * self.P_prop * phi.transpose() + Q_k
        self.P_prop = self.P_pre

        eps = self.q_prop.epsilon()
        self.q_est = self.q_prop + 0.5 * eps * w * dt
        self.q_est.normalize()
        self.q_prop = self.q_est 
    